### PR TITLE
CRD k8s-workload-registrar: Use SPIRE to manage manage webhook certs

### PR DESCRIPTION
**Pull request by faisal-memon**
**Affected functionality**
CRD k8s-workload-registrar

**Description of change**
The change greatly simplifies deployment of the CRD k8s-workload-registrar by using SPIRE certificates for the validating webhook, instead of user-provided. This is managed by the spiffe-helper package. Before the webhook is set up on the k8s-workload-registrar, the spiffe-helper package gets the certs from the SPIRE Server and saves them to disk. Webhooks need certs to be stored on disk. The spiffe-helper package also rotates the CABundle in the ValidatingWebhookConfiguration as needed. Once the certs are setup, the webhook is started.

A quick start has been added that shows the CRD k8s-workload-registrar is now a very easy two step deployment process.

**Caveats**
~~There is a small window during bootstrapping where we can receive reconcile events but the webhook is not started. This is to allow the certificate the k8s-workload-registar uses for the webhook to be created.~~ Fixed with  https://github.com/spiffe/spire/pull/1940/commits/c3c88026b93364b21be6a6d0b4a13de921233b6c

**Dependent changes**
- https://github.com/spiffe/spiffe-helper/issues/5: Updates spiffe-helper to use go-spiffe/v2
- Additional changes in spiffe-helper to rotate webhook CABundles, dependent on the above PR: https://github.com/faisal-memon/spiffe-helper/tree/webhook

### CODE REVIEW

File: support/k8s/k8s-workload-registrar/config_crd.go [hunk 1]
```diff
@@ -149,6 +177,103 @@ func (c *CRDMode) Run(ctx context.Context) error {
...
+	if err != nil {
+		return err
+	}
+
+	log.Info("Setting up SPIFFE sidecar")
+	err = os.MkdirAll(webhookCertDir, 0700)
+	if err != nil {
+		return err
+	}
+	spiffeSidecar, err := sidecar.NewSidecar(&sidecar.Config{
```
Comment (azdagron) [inline support/k8s/k8s-workload-registrar/config_crd.go:217]: Is this fork somehow restarting the webhook when the certificates are updated?

**Review by azdagron — COMMENTED** [review]
Thanks for opening this @faisal-memon. I'd hoped to get the time today to go over this in detail but that hasn't been the case. I'm hoping I'll have the time tomorrow.

From a high level though, I'm a little concerned about pulling in a spiffe-helper dependency, particularly with the kubernetes-specific features you seem to have added in your fork (which are probably not general enough to be appropriate to merge into the upstream spiffe-helper repo).

Can you share details on how the webhook picks up on the changes to rotated certificate/bundle information? Is spiffe-helper overkill here? Can this be pretty trivially done using the go-spiffe/v2 WorkloadAPI client code directly (e.g.  `workloadapi.WatchX509Context`)?

**Comment by faisal-memon** [comment]
> Thanks for opening this @faisal-memon. I'd hoped to get the time today to go over this in detail but that hasn't been the case. I'm hoping I'll have the time tomorrow.
> 
> From a high level though, I'm a little concerned about pulling in a spiffe-helper dependency, particularly with the kubernetes-specific features you seem to have added in your fork (which are probably not general enough to be appropriate to merge into the upstream spiffe-helper repo).
> 
> Can you share details on how the webhook picks up on the changes to rotated certificate/bundle information? Is spiffe-helper overkill here? Can this be pretty trivially done using the go-spiffe/v2 WorkloadAPI client code directly (e.g. `workloadapi.WatchX509Context`)?

Hi Andrew,

Thanks for taking a look. There's two parts to the webhook:
1. The cert and key which live on disk in the target pod. If you update the cert/key on disk, the new version is picked up automatically. There is a [cert watcher](https://github.com/kubernetes-sigs/controller-runtime/blob/master/pkg/webhook/internal/certwatcher/certwatcher.go) that looks for the files being updated.
2. The CABundle in the ValidatingWebookConfiguration. Doing an Update operation will trigger the new CABundle to be used.

The reason to use the spiffe-helper package is because the certificates need to live on disk, and that package has all the code to take in the bundle from the API and write them out to disk. If we call the workload api directly then we'd have to duplicate all that code.

The spiffe-helper pkg seemed to me a logical place to add the code to rotate the CABundle in the ValidatingWebookConfiguration. Then it becomes a nice way to easily manage webhook certificates. We can also create a new pkg or plugin to manage webhooks if you think that will work better. Regardless, I think it would provide value to the community to have a solution for managing webhook certificates with SPIRE.

**Comment by azdagron** [comment]
We all talked offline. @faisal-memon is going to remove the spiffe-helper dependency.

**Comment by faisal-memon** [comment]
Closing for now. Will revisit after https://github.com/spiffe/spire/pull/2022 has been completed

**Pull request closed without merging**